### ISSUE TITLE
GroupedBatchSampler related bug in vision/references/detection/train.py 

### ISSUE THREAD

**Issue by erickim555** [pytorch/vision#2884]
I strongly suspect that there is a bug in the detection trainer code that uses `GroupedBatchSampler` to group images by aspect ratio.

```
if args.distributed:
    train_sampler = torch.utils.data.distributed.DistributedSampler(dataset)
    test_sampler = torch.utils.data.distributed.DistributedSampler(dataset_test)
else:
    train_sampler = torch.utils.data.RandomSampler(dataset)
    test_sampler = torch.utils.data.SequentialSampler(dataset_test)

if args.aspect_ratio_group_factor >= 0:
    group_ids = create_aspect_ratio_groups(dataset, k=args.aspect_ratio_group_factor)
    train_batch_sampler = GroupedBatchSampler(train_sampler, group_ids, args.batch_size)
```
https://github.com/pytorch/vision/blob/cffac640d703196ea9a369166fa8ae587cb5e64d/references/detection/train.py#L80

Due to the random shuffle done by `DistributedSampler` and `RandomSampler`, there is an inconsistency between `train_sampler` and `group_ids`. Specifically: `group_ids` is with respect to the original dataset order (as dictated by `dataset`), but `GroupedBatchSampler` will index into `group_ids` using the indices output by `train_sampler`, eg:

```
def __iter__(self):
    buffer_per_group = defaultdict(list)
    samples_per_group = defaultdict(list)

    num_batches = 0
    for idx in self.sampler:
        group_id = self.group_ids[idx]
```
https://github.com/pytorch/vision/blob/cffac640d703196ea9a369166fa8ae587cb5e64d/references/detection/group_by_aspect_ratio.py#L53

The impact is: `GroupedBatchSampler` will use retrieve the wrong aspect ratios when attempting to batch images with the same aspect ratio together, resulting in batches that are sub-optimally aspect-ratio balanced.

If my understanding is correct, then: to fix this, we'd need to change the `train.py` to ensure that the `train_sampler` and `group_ids` are consistent.
I haven't yet had the time to write a small, contained test case that demonstrates the bug, but just in case I'll create this issue while it's on my mind.

**Comment by erickim555** [comment]
Nevermind, this is correct, I misunderstood things. Sorry about that!
